core/domain/user_services.py

```diff
@@ -295,6 +298,75 @@ def require_valid_username(cls, username):
                         'This username is not available.')
 
 
+class UserAuth(python_utils.OBJECT):
+    """Value object representing a user's authentication information.
+
+    Attributes:
+        user_id: str. The unique ID of the user.
+        gae_id: str. The ID of the user retrieved from GAE.
+        pin: str or None. The PIN of the user's profile for android.
+        parent_user_id: str or None. User id of the full user to which a
+            profile user is associated with. None for full users.
+    """
+
+    def __init__(
+            self, user_id, gae_id, pin=None, parent_user_id=None,
+            deleted=False):
+        """Constructs a UserAuth domain object.
+
+        Args:
+            user_id: str. The unique ID of the user.
+            gae_id: str. The ID of the user retrieved from GAE.
+            pin: str or None. The PIN of the user's profile for android.
+            parent_user_id: str or None. User id of the full user to which a
+                profile user is associated with. None for full users.
+            deleted: bool. Whether the user has requested removal of their
+                account.
+        """
+        self.user_id = user_id
+        self.gae_id = gae_id
+        self.pin = pin
+        self.parent_user_id = parent_user_id
+        self.deleted = deleted
+
+    def validate(self):
+        """Checks that user_id, gae_id and email fields of this UserAuth domain
+        object are valid.
+
+        Raises:
+            ValidationError. The user_id is not str.
+            ValidationError. The gae_id is not str.
+            ValidationError. The pin is not str.
+            ValidationError. The parent_user_id is not str.
+        """
+        if not isinstance(self.user_id, python_utils.BASESTRING):
+            raise utils.ValidationError(
+                'Expected user_id to be a string, received %s' % self.user_id)
+        if not self.user_id:
+            raise utils.ValidationError('No user id specified.')
+        if not is_user_id_correct(self.user_id):
+            raise utils.ValidationError('The user ID is in a wrong format.')
+
+        if (self.gae_id is not None and
+                not isinstance(self.gae_id, python_utils.BASESTRING)):
+            raise utils.ValidationError(
+                'Expected gae_id to be a string, received %s' %
+                self.gae_id
+            )
+
+        if (self.pin is not None and
+                not isinstance(self.gae_id, python_utils.BASESTRING)):
+            raise utils.ValidationError(
+                'Expected PIN to be a string, received %s' %
+                self.pin
+            )
+
+        if (self.parent_user_id is not None and
```

Comment:
We should also validate that:
- PIN and parent ID should always be defined together
- GAE ID and parent ID should never both be present

Reply:
Regarding
> - PIN and parent ID should always be defined together.

I think this is not True. Consider a profile which has a parent id but not a PIN set yet or a full user on android with Parent ID as None but having a defined PIN - both are possibile I believe.

Reply:
The latter is allowed, you’re correct. I don’t think the former is currently allowed, but I suppose that doesn’t mean we should prohibit it here.